Ability to log in with Facebook

davidboll opened the pull request:
#### PT Story: https://www.pivotaltracker.com/story/show/155516301
#### Description
Adds "Login with facebook"

* Sign-up/login with facebook

#### What I have learned working on this feature: 
* Working with OmniAuth

#### Screenshots:
<img width="557" alt="skarmavbild 2018-03-02 kl 11 00 01" src="https://user-images.githubusercontent.com/25644242/36893588-e5eec7bc-1e08-11e8-9143-92d0090aaf71.png">
<img width="1077" alt="skarmavbild 2018-03-02 kl 13 05 55" src="https://user-images.githubusercontent.com/25644242/36898452-2e9eafd8-1e1b-11e8-8d3b-8d0c1abb9f87.png">

davidboll committed:
Adds initial setup with feature file and steps

davidboll committed:
Adds omniauth-facebook to Gemfile

davidboll committed:
Adds omniauth_callback-controller
Adds migration for omniath to users

davidboll committed:
Changes scope from "public_profile" to "email" in config/initializers/devise.rb
Removes unnecessary step from feature-file
Adds facebook logic to navbar
Minor general beutifying of code

davidboll committed:
Adds test for checking fb-columns in DB-table users

davidboll committed:
Adds "Sad path" to cucumber-test
Removes notice from facebook-failure-method in OmniAuth-controller

davidboll committed:
Removes "Sad Path" from cucumber-test

tochman commented on app/models/user.rb:
```diff
@@ -2,5 +2,27 @@ class User < ApplicationRecord
+       #user.image = auth.info.image # assuming the user model has an image
+       # If you are using confirmable and the provider(s) you use validate emails,
+       # uncomment the line below to skip the confirmation emails.
+       # user.skip_confirmation!
```
Please remove comments

oliverochman commented on spec/models/user_spec.rb:
```diff
@@ -13,7 +13,12 @@
+  describe 'db table with added fb-login' do
+    it {is_expected.to have_db_column :uid}
+    it {is_expected.to have_db_column :provider}
+  end
```
You don't need a separate describe block for these tests. I would add them to the existing describe block for the database columns.

davidboll replied:
Tnx! Will change this 👍

davidboll committed:
Adds sad path to test
Combines test of db-columns to one describe block

davidboll committed:
removes "Show me the page" from test

tochman commented on Gemfile:
```diff
@@ -5,32 +5,34 @@ git_source(:github) do |repo_name|
-gem 'rails', '~> 5.1.5'
+gem 'coffee-rails', '~> 4.2'
+gem 'devise'
+gem 'jbuilder', '~> 2.5'
```
This one can go

davidboll replied:
To be clear, you want me to remove which gem? Not sure I want to remove 'coffe-rails' since it broke everything last time, remember?

tochman replied:
I mean `jbuilder`

tochman commented on app/views/layouts/_nav.html.erb:
```diff
@@ -6,6 +6,12 @@
+      <li><% unless current_user %>
+            <%= link_to "Sign in with Facebook", user_facebook_omniauth_authorize_path %>
+          <% else %>
+            <%= current_user[:name] %>
```
I usually use `current_user.name` rather than the hash value markup. Not a blocker but a bit inconsistent.

davidboll replied:
I agree and I'll fix it 👍

tochman commented on config/routes.rb:
```diff
@@ -1,7 +1,7 @@
   get 'expenses/create'
 
-  devise_for :users
+  devise_for :users, controllers: { omniauth_callbacks: 'users/omniauth_callbacks' }
```
I'm not convinced that having this controller in a folder named `users` is the best way to go. I would move it out, rename the controller and change the pointer here in `routes.rb`

davidboll replied:
Ok, I'll give this a go 👍

davidboll committed:
Removes comments from user-model
Refactors "current_user[:name]" to "current_user.name"
Moves OmniAuth-controller out of users-folder in to main controllers-folder

tochman commented on features/step_definitions/authenticated_user_steps.rb:
```diff
@@ -37,6 +37,23 @@
+
+Then("I should be redirected to the landing page") do
+  expect(page.current_path).to eq root_path
+end
```
You already have the following step def:
```
Then("I should be on the {string} page") do |path_name|
  expect(page.current_path).to eq page_path(path_name)
end
```
Why not use that one here?

davidboll replied:
Good point! I'll fix it!

tochman commented on features/step_definitions/authenticated_user_steps.rb:
```diff
@@ -37,6 +37,23 @@
+
+Given("I visit the site") do
+  visit root_path
+end
```
You have this step def already:
```
Given("I am on the {string} page") do \|path_name\|
  visit page_path(path_name)
end
```
Get into the habit of re-using your own steps. Even if I do something else in my demo. Okay?

davidboll replied:
Do as I say, not as I do.. Got it ;)

davidboll committed:
Removes "jbuilder"-gem
Removes unnecessary steps from "steps" and edits step definitions in feature-file to reuse pre-defined steps

SophieSjoberg commented on features/user_can_login_with_facebook.feature:
```diff
@@ -0,0 +1,16 @@
+
+  Scenario: Visitor fails to sign up with Facebook OmniAuth
+    Given I am on the "Landing" page
+    And I have incorrect credetialls
```
Fix spelling on credentials ;)

davidboll replied:
Tnx, I noticed that, but wasn't sure I wanted to push a commit just for a typo :) But yeah, a push on the way ;)

SophieSjoberg replied:
Do you want to merge typos? I think not :)

SophieSjoberg commented on features/step_definitions/authenticated_user_steps.rb:
```diff
@@ -37,6 +37,15 @@
   visit page_path(path_name)
 end
 
+Given("I have incorrect credetialls") do
```
Spelling

davidboll committed:
Fixes typo

SophieSjoberg reviewed (approved):
LGTM
